File: kw_cursor/Assets/Scripts/Import/E5Archive.cs
using System;
using System.Collections.Generic;

namespace KwCursor
{
    /// <summary>
    /// Ls11/Ls12 封包读取器（曹操传/英杰传系列通用容器）。
    /// 格式为社区公开逆向成果（轩辕春秋 van/Maxwell 等），此实现为本项目原创代码。
    ///
    /// 结构：
    ///   0x000  "Ls11"/"Ls12" + 12 字节填充
    ///   0x010  256 字节字典
    ///   0x110  条目表：每条 12 字节（大端 u32：压缩长度、原始长度、偏移），压缩长度 0 结束；
    ///          压缩长度 == 原始长度 表示未压缩存储
    /// 解压：MSB 位流。变长码 = 前缀（读 1 直到 0）+ 等长补码；
    ///   码值 &lt; 0x100 输出 dictionary[码值]；
    ///   码值 &gt;= 0x100 为回溯复制：偏移 = 码值-0x100，长度 = 下一码值+3。
    /// </summary>
    public class E5Archive
    {
        public struct Entry
        {
            public int Compressed;
            public int Original;
            public int Offset;
        }

        byte[] data;
        readonly byte[] dict = new byte[256];
        readonly List<Entry> entries = new List<Entry>();

        public int Count { get { return entries.Count; } }
        public Entry GetEntry(int i) { return entries[i]; }

        public static E5Archive Parse(byte[] bytes)
        {
            if (bytes == null || bytes.Length < 0x110 + 12) return null;
            if (bytes[0] != (byte)'L' || bytes[1] != (byte)'s' || bytes[2] != (byte)'1') return null;
            if (bytes[3] != (byte)'1' && bytes[3] != (byte)'2') return null;

            var a = new E5Archive();
            a.data = bytes;
            Array.Copy(bytes, 0x10, a.dict, 0, 256);

            int off = 0x110;
            while (off + 12 <= bytes.Length)
            {
                int comp = ReadBE(bytes, off);
                int orig = ReadBE(bytes, off + 4);
                int pos = ReadBE(bytes, off + 8);
                off += 12;
                if (comp == 0) break;
                if (orig < 0 || pos < 0 || pos + Math.Max(comp, 0) > bytes.Length) return null;
                a.entries.Add(new Entry { Compressed = comp, Original = orig, Offset = pos });
            }
            return a;
        }

        static int ReadBE(byte[] b, int i)
        {
            return (b[i] << 24) | (b[i + 1] << 16) | (b[i + 2] << 8) | b[i + 3];
        }

        public byte[] Extract(int index)
        {
            Entry e = entries[index];
            var outBuf = new byte[e.Original];
            if (e.Compressed == e.Original)
            {
                Array.Copy(data, e.Offset, outBuf, 0, e.Original);
                return outBuf;
            }

            int pos = e.Offset;
            int curByte = 0;
            int nbits = 0;
            int produced = 0;

            int Bit()
            {
                if (nbits == 0)
                {
                    curByte = data[pos++];
                    nbits = 8;
                }
                int b = (curByte >> 7) & 1;
                curByte = (curByte << 1) & 0xFF;
                nbits--;
                return b;
            }

            int Bits(int n)
            {
                int v = 0;
                for (int i = 0; i < n; i++) v = (v << 1) | Bit();
                return v;
            }

            int Code()
            {
                int c1 = 0, n = 0, b;
                do
                {
                    b = Bit();
                    c1 = (c1 << 1) | b;
                    n++;
                } while (b != 0);
                return c1 + Bits(n);
            }

            while (produced < e.Original)
            {
                int code = Code();
                if (code < 0x100)
                {
                    outBuf[produced++] = dict[code];
                    continue;
                }
                int back = code - 0x100;
                if (back > produced) throw new InvalidOperationException("Ls12 数据损坏：回溯越界");
                int count = Code() + 3;
                for (int i = 0; i < count && produced < e.Original; i++)
                {
                    outBuf[produced] = outBuf[produced - back];
                    produced++;
                }
            }
            return outBuf;
        }
    }
}

File: kw_geminid/src/core/e5/E5Archive.ts
/**
 * Ls11 / Ls12 封包解压缩器（TypeScript 原生实现）
 * 适用于《三国志曹操传》Ekd5 引擎各类 .e5 文件
 */

class BitReader {
  private data: Uint8Array;
  private pos: number;
  private cur: number = 0;
  private nbits: number = 0;

  constructor(data: Uint8Array, offset: number) {
    this.data = data;
    this.pos = offset;
  }

  public bit(): number {
    if (this.nbits === 0) {
      this.cur = this.data[this.pos++];
      this.nbits = 8;
    }
    const b = (this.cur >> 7) & 1;
    this.cur = (this.cur << 1) & 0xff;
    this.nbits--;
    return b;
  }

  public bits(n: number): number {
    let v = 0;
    for (let i = 0; i < n; i++) {
      v = (v << 1) | this.bit();
    }
    return v;
  }
}

function readCode(br: BitReader): number {
  let code1 = 0;
  let n = 0;
  while (true) {
    const b = br.bit();
    code1 = (code1 << 1) | b;
    n++;
    if (b === 0) break;
  }
  const code2 = br.bits(n);
  return code1 + code2;
}

export interface E5Entry {
  compressed: number;
  original: number;
  offset: number;
}

export class E5Archive {
  public magic: string;
  public entries: E5Entry[] = [];
  private data: Uint8Array;
  private dict: Uint8Array = new Uint8Array(256);

  constructor(data: Uint8Array) {
    this.data = data;
    const magicChars = String.fromCharCode(data[0], data[1], data[2], data[3]);
    if (magicChars !== 'Ls11' && magicChars !== 'Ls12') {
      throw new Error(`Invalid magic: ${magicChars}`);
    }
    this.magic = magicChars;
    this.dict.set(data.subarray(0x10, 0x110));

    const view = new DataView(data.buffer, data.byteOffset, data.byteLength);
    let off = 0x110;
    while (off + 12 <= data.byteLength) {
      const comp = view.getUint32(off, false);
      const orig = view.getUint32(off + 4, false);
      const pos = view.getUint32(off + 8, false);
      off += 12;
      if (comp === 0) break;
      this.entries.push({ compressed: comp, original: orig, offset: pos });
    }
  }

  public get count(): number {
    return this.entries.length;
  }

  public extract(index: number): Uint8Array {
    if (index < 0 || index >= this.entries.length) {
      throw new Error(`Entry index out of bounds: ${index}`);
    }
    const e = this.entries[index];
    if (e.compressed === e.original) {
      return this.data.subarray(e.offset, e.offset + e.original);
    }

    const out = new Uint8Array(e.original);
    let produced = 0;
    const br = new BitReader(this.data, e.offset);

    while (produced < e.original) {
      const code = readCode(br);
      if (code < 0x100) {
        out[produced++] = this.dict[code];
        continue;
      }
      const back = code - 0x100;
      if (back > produced) {
        throw new Error(`Ls12 corrupt back pointer: ${back} > ${produced}`);
      }
      const count = readCode(br) + 3;
      for (let i = 0; i < count && produced < e.original; i++) {
        out[produced] = out[produced - back];
        produced++;
      }
    }
    return out;
  }
}
